### PR TITLE
tsdb index gateway

### PR DESCRIPTION
**Pull request by sandeepsukhani**
**What this PR does / why we need it**:
1. enable new grpc calls on index gateway and add GetSeries call
2. Use index gateway client when enabled for boltdb-shipper index store.

**Special notes for your reviewer**:
* Please do not merge this PR until we are ready to enable new grpc calls or we add a flag to control it since we had disabled them until we can improve index gateway performance.
* We rely on boltdb-shipper config for now to initialize the index gateway client. We will have to discuss and decide how we want to proceed with tsdb config. I feel that we should share the same config to avoid complexities.

### CODE REVIEW

File: pkg/storage/store.go [hunk 1]
```diff
@@ -200,58 +198,34 @@ func shouldUseBoltDBIndexGatewayClient(cfg Config) bool {
...
-			tsdbMetrics,
-			tsdbManager,
-		)
-		if err := headManager.Start(); err != nil {
-			return nil, nil, nil, err
-		}
-		idx := tsdb.NewIndexClient(headManager, p)
-		writer := tsdb.NewChunkWriter(f, p, headManager)
+		// ToDo(Sandeep): Avoid initializing writer when in read only mode
+		writer, idx, err := tsdb.NewStore(s.cfg.TSDBShipperConfig, p, f, objectClient, s.limits, indexClientReg)
```
Comment (owen-d) [inline pkg/storage/store.go:212]: I don't see `tsdb.NewStore` implemented yet, is that right? I'm guessing we'll just copy this over to the new function (we can also add a check on whether to enable the head manager as well).

**Comment by sandeepsukhani** [comment]
> Did a quick review; this is looking good!
> 
> > Please do not merge this PR until we are ready to enable new grpc calls or we add a flag to control it since we had disabled them until we can improve index gateway performance.
> 
> Instead, can we create two clients, one which uses the new form and one which uses the old form? Then we can enable the new form when tsdb is enabled, but keep the old form elsewhere.

We already have a fallback mechanism for backwards compatibility. I think we should add a flag to reuse most of the code then.

File: pkg/storage/store.go [hunk 2]
```diff
@@ -200,58 +198,34 @@ func shouldUseBoltDBIndexGatewayClient(cfg Config) bool {
...
 
-		// TODO(owen-d): add TSDB index-gateway support
+		// ToDo(Sandeep): Refactor code to not use boltdb-shipper index gateway client config
+		if shouldUseBoltDBIndexGatewayClient(s.cfg) {
```
[the remark this answers is not in the document]
Comment (sandeepsukhani) [reply pkg/storage/store.go:215]: This should all be cleaned up once I get to migrate `boltdb-shipper` to `index-shipper`.

File: pkg/storage/stores/composite_store_entry.go [hunk 3]
```diff
@@ -25,17 +25,6 @@ type StoreLimits interface {
 	MaxQueryLength(userID string) time.Duration
 }
 
-type Index interface {
```
[the remark this answers is not in the document]
Comment (sandeepsukhani) [reply pkg/storage/stores/composite_store_entry.go:28]: I wanted to pass an interface to [series.NewIndexGatewayClientStore](https://github.com/grafana/loki/pull/6158/files#diff-22f496b77143a9ff61e9e711e43208e2e9258130e04c672a1c3873602bd937aaR29) without which it was not possible to pass tsdb index client.

File: pkg/storage/store.go (hunk 1 above)
Comment (sandeepsukhani) [reply pkg/storage/store.go:212]: Sorry, I missed pushing the file. I have pushed it yesterday.

**Comment by sandeepsukhani** [comment]
> > Did a quick review; this is looking good!
> > > Please do not merge this PR until we are ready to enable new grpc calls or we add a flag to control it since we had disabled them until we can improve index gateway performance.
> > 
> > 
> > Instead, can we create two clients, one which uses the new form and one which uses the old form? Then we can enable the new form when tsdb is enabled, but keep the old form elsewhere.
> 
> We already have a fallback mechanism for backwards compatibility. I think we should add a flag to reuse most of the code then.

I have temporarily disabled using `IndexGatewayClientStore` for `boltdb-shipper` since that is what tries the new storage grpc calls on index gateway. Index gateway should continue to work as is now since we create index gateway client instance [here](https://github.com/grafana/loki/blob/e2fdc3437b38c67d817028ce4f002a6f20ca9e1e/pkg/storage/factory.go#L159) anyways.

**Review by owen-d — APPROVED** [review]

**Pull request merged**